README.md
# ptyng


CHANGELOG.md
# Changelog

## Version 0.3.3

 > Release date: 2019-01-08

&emsp; If attempt to kill child w/ `SIGKILL`failed, use `SIGTERM` instead.

## Version 0.3.2

 > Release date: 2019-01-07

&emsp; Kill child process (and its children) in a graceful way (to avoid orphan).

## Version 0.3.1.post1

 > Release date: 2019-01-05

&emsp; Fixed bugs in `setup.py`.

## Version 0.3.1

 > Release date: 2019-01-04

&emsp; Fixed bugs when killing spawned process.

## Version 0.3.0.post2

 > Release date: 2018-12-12

&emsp; There were no new changes in version 0.3.0.post2. But minor bugs fixed in `setup.py` with encoding errors.

## Version 0.3.0.post1

 > Release date: 2018-12-08

&emsp; There are no updates in version 0.3.0.post1.

## Version 0.3.0

 > Release date: 2018-12-08

- no changes
- changed license & updated misc files

## Version 0.2.1.post2

 > Release date: 2018-11-28

&emsp; Fixed distribution errors, and uploaded wheels for CPython 3.4/3.5.

## Version 0.2.1.post1

 > Release date: 2018-11-25

&emsp; There are no updates in version 0.2.1.post1.

## Version 0.2.1

 > Release date: 2018-11-24

&emsp; Added `env` argument to `spawn`.

## Version 0.2.0.post4

 > Release date: 2018-11-19

&emsp; Updated distribution script.

## Version 0.2.0.post3

 > Release date: 2018-11-15

&emsp; There are no updates in version 0.2.0.post3.

## Version 0.2.0.post2

 > Release date: 2018-11-15

&emsp; There are no updates in version 0.2.0.post2.

## Version 0.2.0.post1

 > Release date: 2018-11-13

&emsp; There are no updates in version 0.2.0.post1.

## Version 0.2.0

 > Release date: 2018-10-31

&emsp; Fixed minor bugs with `zombie` determination.

## Version 0.1.3

 > Release date: 2018-10-11

&emsp; Fixed minor bugs with `timeout` when cancelled.

## Version 0.1.2

 > Release date: 2018-09-27

&emsp; Added `timeout` argument to `ptyng.spawn`.

## Version 0.1.1.post3

 > Release date: 2018-09-26

&emsp; There are no updates in version 0.1.1.post3.

## Version 0.1.1.post2

 > Release date: 2018-09-26

&emsp; There are no updates in version 0.1.1.post2.

## Version 0.1.1.post1

 > Release date: 2018-09-26

&emsp; There are no updates in version 0.1.1.post1.

## Version 0.1.1

 > Release date: 2018-09-25

&emsp; Introduced `psutil` for cross-platform compatibility.

## Version 0.1.0.post2

 > Release date: 2018-09-24

&emsp; There are no updates in version 0.1.0.post2.

## Version 0.1.0.post1

 > Release date: 2018-09-24

&emsp; There are no updates in version 0.1.0.post1.

## Version 0.1.0

 > Release date: 2018-09-24

&emsp; Fork of `pty` aiming to enhance performance of the full `stdlib` Python API.


test.py
# import io
# import os
# import pty
# import subprocess

# class IOBase(io.IOBase):

#     def fileno(self):
#         return self._fileno

#     def __init__(self, fileno, *args, **kwargs):
#         self._fileno = fileno
#         super().__init__(*args, **kwargs)

# # parent, child = pty.openpty()
# # print('ptyed!') ###
# # # subprocess.run(['brew', 'info'], stdout=IOBase(child))
# # # subprocess.run(['brew', 'info'], stdout=IOBase(child))
# # # subprocess.run(['curl https://httpbin.org/ip; sleep 5'], shell=True, stdout=IOBase(child), stderr=subprocess.STDOUT, timeout=60)
# # subprocess.run(['brew', 'info', 'python'], stdout=IOBase(child), stderr=subprocess.STDOUT, timeout=60)
# # print('subprocessed!') ###
# # print(os.read(parent, 1024))



# pid, fd = pty.fork()
# if pid == 0:
#     print(fd)
#     subprocess.run(['npm', 'help'])
# else:
#     print(fd)
#     print(os.read(fd, 1024))
#     print(os.wait4(pid, os.WEXITED))


import argparse
import os
import pty
import sys
import time

parser = argparse.ArgumentParser()
parser.add_argument('-a', dest='append', action='store_true')
parser.add_argument('-p', dest='use_python', action='store_true')
parser.add_argument('filename', nargs='?', default='typescript')
options = parser.parse_args()

shell = sys.executable if options.use_python else os.environ.get('SHELL', 'sh')
filename = options.filename
mode = 'ab' if options.append else 'wb'

with open(filename, mode) as script:
    def read(fd):
        data = os.read(fd, 1024)
        script.write(data)
        return data

    print('Script started, file is', filename)
    script.write(('Script started on %s\n' % time.asctime()).encode())

    # pty.spawn(shell, read)
    pty.spawn(['brew', 'info', 'python'], read)

    script.write(('Script done on %s\n' % time.asctime()).encode())
    print('Script done, file is', filename)


make.sh
#!/usr/bin/env bash

# print a trace of simple commands
set -x

# prepare for PyPI distribution
rm -rf build 2> /dev/null
mkdir eggs \
      sdist \
      wheels 2> /dev/null
mv -f dist/*.egg eggs/ 2> /dev/null
mv -f dist/*.whl wheels/ 2> /dev/null
mv -f dist/*.tar.gz sdist/ 2> /dev/null
rm -rf dist 2> /dev/null

# fetch platform spec
platform=$( python3 -c "import distutils.util; print(distutils.util.get_platform().replace('-', '_').replace('.', '_'))" )

# make distribution
python3.7 setup.py sdist bdist_egg bdist_wheel --plat-name="${platform}" --python-tag='cp37'
python3.6 setup.py bdist_egg bdist_wheel --plat-name="${platform}" --python-tag='cp36'
python3.5 setup.py bdist_egg bdist_wheel --plat-name="${platform}" --python-tag='cp35'
python3.4 setup.py bdist_egg bdist_wheel --plat-name="${platform}" --python-tag='cp34'
python2.7 setup.py bdist_egg bdist_wheel --plat-name="${platform}" --python-tag='cp27'
pypy3 setup.py bdist_wheel --plat-name="${platform}" --python-tag='pp35'
pypy setup.py bdist_wheel --plat-name="${platform}" --python-tag='pp27'

# distribute to PyPI and TestPyPI
twine upload dist/* -r pypi --skip-existing
twine upload dist/* -r pypitest --skip-existing

# get version string
version=$( cat setup.py | grep "^__version__" | sed "s/__version__ = '\(.*\)'/\1/" )

# upload to GitHub
git pull && \
git tag "v${version}" && \
git add . && \
if [[ -z "$1" ]] ; then
    git commit -a -S
else
    git commit -a -S -m "$1"
fi && \
git push
ret="$?"
if [[ $ret -ne "0" ]] ; then
    exit $ret
fi

# file new release
go run github.com/aktau/github-release release \
    --user JarryShaw \
    --repo ptyng \
    --tag "v${version}" \
    --name "ptyng v${version}" \
    --description "$1"
ret="$?"
if [[ $ret -ne "0" ]] ; then
    exit $ret
fi

# update maintenance information
# maintainer changelog && \
maintainer contributor && \
maintainer contributing
ret="$?"
if [[ $ret -ne "0" ]] ; then
    exit $ret
fi

# aftermath
git pull && \
git add . && \
git commit -a -S -m "Regular update after distribution" && \
git push


ptyng.py
"""Pseudo terminal utilities."""

# Bugs: No signal handling.  Doesn't set slave termios and window size.
#       Only tested on Linux.
# See:  W. Richard Stevens. 1992.  Advanced Programming in the
#       UNIX Environment.  Chapter 19.
# Author: Steen Lumholt -- with additions by Guido.

import contextlib
import io
import os
import signal
import sys
import tty
import warnings
from select import select

try:
    import threading
except ImportError:
    import dummy_threading as threading

try:
    from shutil import which
except ImportError:
    from backports.shutil_which import which

if sys.version_info[:2] < (3, 4):
    import subprocess32 as subprocess
else:
    import subprocess

try:
    import psutil
    TIMEOUT = 1
except ImportError:
    psutil = None
    PS_PATH = which('ps')
    PGREP_PATH = which('pgrep')
    TIMEOUT = None if PS_PATH is None else 1

__all__ = ["openpty", "fork", "spawn"]

STDIN_FILENO = 0
STDOUT_FILENO = 1
STDERR_FILENO = 2

CHILD = 0


def openpty():
    """openpty() -> (master_fd, slave_fd)
    Open a pty master/slave pair, using os.openpty() if possible."""

    try:
        return os.openpty()
    except (AttributeError, OSError):
        pass
    master_fd, slave_name = _open_terminal()
    slave_fd = slave_open(slave_name)
    return master_fd, slave_fd


def master_open():
    """master_open() -> (master_fd, slave_name)
    Open a pty master and return the fd, and the filename of the slave end.
    Deprecated, use openpty() instead."""

    try:
        master_fd, slave_fd = os.openpty()
    except (AttributeError, OSError):
        pass
    else:
        slave_name = os.ttyname(slave_fd)
        os.close(slave_fd)
        return master_fd, slave_name

    return _open_terminal()


def _open_terminal():
    """Open pty master and return (master_fd, tty_name)."""
    for x in 'pqrstuvwxyzPQRST':
        for y in '0123456789abcdef':
            pty_name = '/dev/pty' + x + y
            try:
                fd = os.open(pty_name, os.O_RDWR)
            except OSError:
                continue
            return (fd, '/dev/tty' + x + y)
    raise OSError('out of pty devices')


def slave_open(tty_name):
    """slave_open(tty_name) -> slave_fd
    Open the pty slave and acquire the controlling terminal, returning
    opened filedescriptor.
    Deprecated, use openpty() instead."""

    result = os.open(tty_name, os.O_RDWR)
    try:
        from fcntl import ioctl, I_PUSH
    except ImportError:
        return result
    try:
        ioctl(result, I_PUSH, "ptem")
        ioctl(result, I_PUSH, "ldterm")
    except OSError:
        pass
    return result


def fork():
    """fork() -> (pid, master_fd)
    Fork and make the child a session leader with a controlling terminal."""

    try:
        pid, fd = os.forkpty()
    except (AttributeError, OSError):
        pass
    else:
        if pid == CHILD:
            try:
                os.setsid()
            except OSError:
                # os.forkpty() already set us session leader
                pass
        return pid, fd

    master_fd, slave_fd = openpty()
    pid = os.fork()
    if pid == CHILD:
        # Establish a new session.
        os.setsid()
        os.close(master_fd)

        # Slave becomes stdin/stdout/stderr of child.
        os.dup2(slave_fd, STDIN_FILENO)
        os.dup2(slave_fd, STDOUT_FILENO)
        os.dup2(slave_fd, STDERR_FILENO)
        if (slave_fd > STDERR_FILENO):
            os.close(slave_fd)

        # Explicitly open the tty to make it become a controlling tty.
        tmp_fd = os.open(os.ttyname(STDOUT_FILENO), os.O_RDWR)
        os.close(tmp_fd)
    else:
        os.close(slave_fd)

    # Parent and child process.
    return pid, master_fd


def _writen(fd, data):
    """Write all the data to a descriptor."""
    while data:
        n = os.write(fd, data)
        data = data[n:]


def _read(fd):
    """Default read function."""
    return os.read(fd, 1024)


if psutil is None:  # if psutil not installed
    def _is_zombie(pid):
        """Check if pid is in zombie stat."""
        if PS_PATH is None:
            return False
        try:
            proc = subprocess.check_output([PS_PATH, 'axo', 'pid=,stat='])
        except subprocess.CalledProcessError:
            return False
        for line in proc.strip().splitlines():
            _pid, stat = line.strip().decode().split()
            if int(_pid) == pid:
                return ('Z' in stat)
        raise OSError(3, 'No such process')

    def _fetch_child(ppid):
        """Get child processes of a leading process."""
        chld_pid = [ppid]
        if PGREP_PATH is None:
            return chld_pid
        try:
            proc = subprocess.check_output([PGREP_PATH, '-P', str(ppid)])
        except subprocess.CalledProcessError:
            return chld_pid
        for line in proc.strip().splitlines():
            with contextlib.suppress(ValueError):
                chld_pid.extend(_fetch_child(int(line.strip())))
        return chld_pid
else:
    def _is_zombie(pid):
        """Check if pid is in zombie stat."""
        try:
            return (psutil.Process(pid).status() == psutil.STATUS_ZOMBIE)
        except psutil.NoSuchProcess:
            raise OSError(3, 'No such process')

    def _fetch_child(ppid):
        """Get child processes of a leading process."""
        chld_pid = [ppid]
        try:
            for child in psutil.Process(ppid).children(recursive=True):
                chld_pid.append(child.pid)
        except psutil.NoSuchProcess:
            pass
        return chld_pid


def _copy(pid, master_fd, master_read=_read, stdin_read=_read):
    """Parent copy loop.
    Copies
            pty master -> standard output   (master_read)
            standard input -> pty master    (stdin_read)"""
    fds = [master_fd, STDIN_FILENO]
    while True:
        rfds, _, _ = select(fds, [], [], TIMEOUT)
        if _is_zombie(pid):
            raise OSError(5, 'Input/output error')
        if master_fd in rfds:
            data = master_read(master_fd)
            if not data:  # Reached EOF.
                fds.remove(master_fd)
            else:
                os.write(STDOUT_FILENO, data)
        if STDIN_FILENO in rfds:
            data = stdin_read(STDIN_FILENO)
            if not data:
                fds.remove(STDIN_FILENO)
            else:
                _writen(master_fd, data)


def _kill(pid, signal, master_read):
    """Kill a process with a signal."""
    class FileObject(io.IOBase):
        def write(self, data):
            os.write(master_read, bytes(data, 'utf-8', 'replace'))

    file = FileObject(master_read)
    for chld in reversed(_fetch_child(pid)):
        try:
            os.kill(chld, signal)
        except OSError as error:
            with contextlib.suppress(OSError):
                os.kill(chld, signal.SIGTERM)
            message = ('failed to send signal to process %d '
                       'with error message: %s') % (chld, error)
            warnings.showwarning(message, ResourceWarning, __file__, 246, file)


def spawn(argv, master_read=_read, stdin_read=_read, timeout=None, env=None):
    """Create a spawned process."""
    if env is None:
        env = os.environ
    if isinstance(argv, str):
        argv = (argv,)

    pid, master_fd = fork()
    if pid == CHILD:
        os.execvpe(argv[0], argv, env)

    if timeout is not None:
        timer = threading.Timer(timeout, _kill, args=(pid, signal.SIGKILL, master_fd))
        timer.start()

    try:
        mode = tty.tcgetattr(STDIN_FILENO)
        tty.setraw(STDIN_FILENO)
        restore = 1
    except tty.error:    # This is the same as termios.error
        restore = 0

    try:
        _copy(pid, master_fd, master_read, stdin_read)
        # print('copied!')
    except OSError:
        if restore:
            tty.tcsetattr(STDIN_FILENO, tty.TCSAFLUSH, mode)

    os.close(master_fd)
    if timeout is not None:
        timer.cancel()
    return os.waitpid(pid, 0)[1]
